packages/medusa/src/services/gift-card.ts

```diff
@@ -0,0 +1,317 @@
+import RegionService from "./region"
+
+type InjectedDependencies = {
+  manager: EntityManager
+  giftCardRepository: typeof GiftCardRepository
+  giftCardTransactionRepository: typeof GiftCardTransactionRepository
+  regionService: RegionService
+  eventBusService: EventBusService
+}
+/**
+ * Provides layer to manipulate gift cards.
+ */
+class GiftCardService extends TransactionBaseService<GiftCardService> {
+  protected readonly giftCardRepository_: typeof GiftCardRepository
+  protected readonly giftCardTransactionRepo_: typeof GiftCardTransactionRepository
+  protected readonly regionService_: RegionService
+  protected readonly eventBus_: EventBusService
+
+  protected manager_: EntityManager
+  protected transactionManager_: EntityManager | undefined
+
+  static Events = {
+    CREATED: "gift_card.created",
+  }
+
+  constructor({
+    manager,
+    giftCardRepository,
+    giftCardTransactionRepository,
+    regionService,
+    eventBusService,
+  }: InjectedDependencies) {
+    // eslint-disable-next-line prefer-rest-params
+    super(arguments[0])
+
+    this.manager_ = manager
+
+    this.giftCardRepository_ = giftCardRepository
+    this.giftCardTransactionRepo_ = giftCardTransactionRepository
+    this.regionService_ = regionService
+    this.eventBus_ = eventBusService
+  }
+
+  /**
+   * Generates a 16 character gift card code
+   * @return the generated gift card code
+   */
+  generateCode_(): string {
+    const code = [
+      randomize("A0", 4),
+      randomize("A0", 4),
+      randomize("A0", 4),
+      randomize("A0", 4),
+    ].join("-")
+
+    return code
+  }
+
+  /**
+   * @param selector - the query object for find
+   * @param config - the configuration used to find the objects. contains relations, skip, and take.
+   * @return the result of the find operation
+   */
+  async list(
+    selector: QuerySelector<GiftCard> = {},
+    config: FindConfig<GiftCard> = { relations: [], skip: 0, take: 10 }
+  ): Promise<GiftCard[]> {
+    return await this.atomicPhase_(async (manager) => {
+      const giftCardRepo = manager.getCustomRepository(this.giftCardRepository_)
+
+      let q
+      if ("q" in selector) {
+        q = selector.q
+        delete selector.q
+      }
+
+      const query = buildQuery(selector, config)
+
+      const rels = query.relations as Array<keyof GiftCard>
+      delete query.relations
+
+      if (q) {
+        const where = query.where
+        delete where.id
+
+        const raw = await giftCardRepo
+          .createQueryBuilder("gift_card")
+          .leftJoinAndSelect("gift_card.order", "order")
+          .select(["gift_card.id"])
+          .where(where)
+          .andWhere(
+            new Brackets((qb) => {
+              return qb
+                .where(`gift_card.code ILIKE :q`, { q: `%${q}%` })
+                .orWhere(`display_id::varchar(255) ILIKE :dId`, { dId: `${q}` })
+            })
+          )
+          .getMany()
+
+        return giftCardRepo.findWithRelations(
```

Comment:
suggestion : i suggest to replace that and the query above by a method on the repo such as other entities does. Wdyt? The migration can handle a little cleanup at the same time to make the code base more consistent and bug free wdyt?